### ISSUE TITLE
metrics filesystem cleanup can crash go-carbon

### ISSUE THREAD

**Issue by blysik** [go-graphite/go-carbon#174]
version: go-carbon.0.10.0_beta1-1 
os: Red Hat Enterprise Linux Server release 6.4 (Santiago)

Twice now I've crashed go-carbon by removing stale metrics from the filesystem.  The logs don't show any errors.  

I'm sorry that I don't have a lot of information.  I don't have an environment where I can safely try to reproduce it, at the moment.

**Comment by lomik** [comment]
Do you use carbonserver?

**Comment by lomik** [comment]
How do you detect stale metrics? Are they being written during the cleaning process?

**Comment by nisc-acooper** [comment]
We are also noticing a unidentified crash with 0.10.0_beta1-1 and also do filesystem-level cleanup.  Our cleanup process simply looks for old modify time stamps and removes whisper files.   Prior to 0.10 (i think we were on 0.7.0) we did not have any crashes.  We did turn on carbonserver with 0.10 and use carbonapi in front of it.

**Comment by Civil** [comment]
If it crashes, there should be a panic trace or something about that. It would be really useful to obtain that

**Comment by blysik** [comment]
My setup is:

carbon-c-relay -> go-carbon tcp:4003 for writes 
carbon-api -> carbonzipper -> go-carbon (carbonserver) for reads.  

The relay/api layer is 4 VMs, and the go-carbon layer is 3 physical machines using carbon_ch hash.

I was just manually identifying hosts which had gone away, and removing the metrics manually. They were not writing at the time.  (Double checked, and no new whisper files were created.)

**Comment by blysik** [comment]
Checked my logs again, and there's no panic trace.  (I'm running at INFO level logging, so I'd expect something.)

I have about a minute gap in the logs between the last fetch, and then carbonserver startup when I manually restarted the service. (Truncated lines with ...)

```
{"level":"INFO","timestamp":"2017-05-31T11:29:36.689-0700","logger":"access","message":"fetch served","handler":"render","url":"/render/?format=protobuf&from=1496244576.....
{"level":"INFO","timestamp":"2017-05-31T11:30:21.650-0700","logger":"carbonserver","message":"starting carbonserver","whisperData":....
```

**Comment by Civil** [comment]
If panic is unhandeled, it will be printed out in stdout of the daemon

**Comment by blysik** [comment]
Where does stdout go for RHEL6 init.d service go?  I grepped through /var/log/messages, and I didn't see anything.  I'll look more.

**Comment by blysik** [comment]
Looking at the init script, stdout isn't captured anywhere unfortunately.

**Comment by nisc-acooper** [comment]
We are running on centos 7 with the systemd unit file.  I cannot find any trace of a panic in journal output, I am currently attempting to reproduce the crash while watching all log output (journal + go-carbon.log)

**Comment by blysik** [comment]
@nisc-acooper any luck in reproducing?

**Comment by Civil** [comment]
Can you please check if crash is happening in the PR #175 ?

Also in that PR even if there is a bug in the code, it should log it properly.

**Comment by blysik** [comment]
Any chance you could package a beta2 RPM which contains this fix?  (Or post a linux x86_64 binary somewhere.) I don't have a good build environment right now.

**Comment by lomik** [comment]
> Any chance you could package a beta2 RPM which contains this fix? (Or post a linux x86_64 binary somewhere.) I don't have a good build environment right now.

beta2 version has been released

**Comment by blysik** [comment]
I installed beta2, and did a whole bunch of namespace cleanup.  I was not able to trigger the crash.

**Comment by claudio-benfatto** [comment]
This is related to https://github.com/lomik/go-carbon/issues/186 as well

**Comment by lomik** [comment]
> This is related to #186 as well

No, this issue was fixed and released in 0.10.0
